File: spring-data-requery/src/main/java/org/springframework/data/requery/annotation/Query.java
package org.springframework.data.requery.annotation;

import org.springframework.data.annotation.QueryAnnotation;

import java.lang.annotation.*;

/**
 * com.coupang.springframework.data.requery.annotation.Query
 *
 * @author debop
 * @since 18. 6. 6
 */
@Target({ ElementType.METHOD, ElementType.ANNOTATION_TYPE })
@Retention(RetentionPolicy.RUNTIME)
@QueryAnnotation
@Documented
public @interface Query {

    /**
     * Defines the SQL query to be executed when the annotated queryMethod is called.
     */
    String value() default "";

    /**
     * Defines a special count query that shall be used for pagination queries to lookup the total number of elements for
     * a page. If non is configured we will derive the count query from the queryMethod name.
     */
    String countQuery() default "";

}

File: spring-data-requery/src/main/kotlin/org/springframework/data/requery/annotation/Query.kt
package org.springframework.data.requery.annotation

import org.springframework.data.annotation.QueryAnnotation

/**
 * Query
 *
 * @author debop@coupang.com
 * @since 18. 7. 2
 */
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER, AnnotationTarget.ANNOTATION_CLASS)
@Retention(AnnotationRetention.RUNTIME)
@QueryAnnotation
@MustBeDocumented
annotation class Query(
    /**
     * Defines the SQL query to be executed when the annotated queryMethod is called.
     */
    val value: String = "",
    /**
     * Defines a special count query that shall be used for pagination queries to lookup the total number of elements for
     * a page. If non is configured we will derive the count query from the queryMethod name.
     */
    val countQuery: String = ""

)